packages/core/src/factory.ts

```diff
@@ -18,62 +22,71 @@ export function factory<Resource, Input, Output, Config extends IPrismConfig>(
     request: (input: Input, resources: Resource[], c?: Config) => {
       // build the config for this request
       const config = defaults<unknown, Config>(c, defaultConfig);
-      const inputValidations: IPrismDiagnostic[] = [];
 
       return pipe(
         TaskEither.fromEither(components.route({ resources, input })),
-        TaskEither.chain(resource => {
-          if (config.validateRequest) {
-            inputValidations.push(
-              ...components.validateInput({
-                resource,
-                element: input,
-              })
-            );
-          }
-
-          config.checkSecurity
-            ? inputValidations.push(
-                ...pipe(
-                  validateSecurity(input, resource),
-                  map(sec => [sec]),
-                  getOrElse<IPrismDiagnostic[]>(() => [])
-                )
+        TaskEither.chain(resource =>
+          TaskEither.fromEither(
+            pipe(
+              sequenceValidation(
+                config.validateRequest ? components.validateInput({ resource, element: input }) : Either.right(input),
+                config.checkSecurity ? validateSecurity(input, resource) : Either.right(input)
+              ),
+              Either.fold(
+                validations =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations,
+                  }),
+                () =>
+                  Either.right<Error, { resource: Resource; validations: IPrismDiagnostic[] }>({
+                    resource,
+                    validations: [],
+                  })
               )
-            : inputValidations;
-
-          const outputLocator = isProxyConfig(config)
+            )
+          )
+        ),
+        TaskEither.chain(({ resource, validations }) => {
+          const produceOutput = isProxyConfig(config)
             ? components.forward(input, config.upstream.href)
             : TaskEither.fromEither(
                 components.mock({
                   resource,
                   input: {
-                    validations: inputValidations,
+                    validations,
```

Comment:
note to self: both request and security validation results

Reply:
Let me know once you clarify the question so I can look into.

Reply:
it was more of a `note to self`, no question here, I'm good :+1: